walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,120 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  private ArrayList<TimeRange> getMergedTimes(ArrayList<TimeRange> takenTimes) {
+    Collections.sort(takenTimes, TimeRange.ORDER_BY_START);
+    TimeRange previous = takenTimes.get(0);
+    ArrayList<TimeRange> takenTimesMerged = new ArrayList();
+
+    for (TimeRange time : takenTimes) {
+      if (time.overlaps(previous)) {
+        int start = previous.start();
+        int end = Math.max(time.end(), previous.end());
+        previous = TimeRange.fromStartEnd(start, end, false);
+      } else {
+          takenTimesMerged.add(previous);
+          previous = time;
+        }
+    }
+    takenTimesMerged.add(previous);
+    return takenTimesMerged;
+  }
+
+  private ArrayList<TimeRange> getBetweenTimes(ArrayList<TimeRange> takenTimesMerged) {
+    ArrayList<TimeRange> validTimes = new ArrayList();
+    TimeRange firstValid = takenTimesMerged.get(0);
+    TimeRange lastValid = takenTimesMerged.get(takenTimesMerged.size() - 1);
+    int first = firstValid.start();
+    int last = lastValid.end();
+    validTimes.add(TimeRange.fromStartEnd(0, first ,false));
+    validTimes.add(TimeRange.fromStartEnd(last, 1440, false));
+
+    if (takenTimesMerged.size()>1) {
+      for (int i = 0; i < takenTimesMerged.size(); i++) {
+        if (i == takenTimesMerged.size()-1) {
+          break;
+        }
+        TimeRange currentRange = takenTimesMerged.get(i);
+        TimeRange adjRange = takenTimesMerged.get(i+1);
+        int newStart = currentRange.end();
+        int newEnd = adjRange.start();
+        validTimes.add(TimeRange.fromStartEnd(newStart, newEnd, false));
+      }
+    }
+    return validTimes;
   }
+
+  private ArrayList<TimeRange> filterTimes(ArrayList<TimeRange> validTimes, MeetingRequest request) {
+    ArrayList<TimeRange> finalTimes = new ArrayList();
+    for (TimeRange time : validTimes) {
+      if (time.duration() >= request.getDuration()) {
+          finalTimes.add(time);
+      }
+    }
+    Collections.sort(finalTimes, TimeRange.ORDER_BY_START);
```

Comment:
Not that this should be expensive if the list was already in order but I don't think this is needed.

Reply:
For some reason the time slots have to be ordered when returned or the test cases will fail

Reply:
![Screenshot 2020-07-16 at 12 30 31 PM](https://user-images.githubusercontent.com/51325209/87714162-3a78bb00-c760-11ea-9ef5-7a8090c4d722.png)

Reply:
Ooooooh, right this goes back to you adding the first and last time slots before going into the loop above. So yeah if you don't do that then order should be assured and you won't need this sort here.